### ISSUE TITLE
Autocomplete Doesn't Work

### ISSUE THREAD

**Issue by ashes999** [vshaxe/vshaxe#122]
Like [this issue](https://github.com/vshaxe/vshaxe/issues/117), I can't get vscode autocomplete to work. Unlike that issue, though:

- There are no non-ASCII characters in my path
- My project is indeed on the E:\ drive (red herring, I know)
- I cloned @Gama11 's [haxe repository](https://github.com/Gama11/HaxeRepro) and it works properly

I copied over the `build.hxml`, `settings.json`, `launch.json`, and `tasks.json` from his project. Still didn't work. I thought it was my use of libs (`flixel`, `flixel-addons`, and `[turbo](github.com/deengames/turbo)`) but his project also works if I add those to `Build.hxml`. 

What am I missing?

My code is [here](https://github.com/deengames-prototypes/pattern-warrior).

**Comment by Gama11** [comment]
For flixel projects, please use flixel-tools / flixel-templates according to the guide here (use the `configure` command to add a .vscode config to an existing project):

http://haxeflixel.com/documentation/visual-studio-code/

Note that you'll have to open the `code` subfolder of your repo in VSCode, as that's where the `Project.xml` file is located.

**Comment by ashes999** [comment]
Now, this is interesting. This works if I build my project with `flash` as the target, but not with `neko` and `html5` builds in `export`.

Hmm. Anyways, this is resolved, (albeit not ideally in terms of experience) --  thanks for sharing that link and providing the instructions about opening `code` in VSCode.

**Comment by Gama11** [comment]
What doesn't work on neko / html5?

**Comment by ashes999** [comment]
Sorry, bad explanation. Let me try again.

- I followed your instructions and used the flixel command-line (option 3: generate for an existing project) in my project.
- I opened up the `code` folder in VSCode
- I tried autocomplete. No dice. 
- I created a new project following the wiki page, verified it worked, copied over the `.vscode` folder to my project, and tried autocomplete. No dice.
- I ran `lime build flash` from my project root, then tried auto-complete.  *It worked.*

Now, technically, `flash` shouldn't make a difference; `settings.json` contains entries for `html5` and `neko`, both of which existed in my `export` directory at the time that I tried to get auto-complete to work (without `flash`).

But, **only by compiling my project for Flash, did I get auto-complete to work in VSCode.** Without that, despite following the instructions and building my project in neko/html5, autocomplete didn't work.

Is this a bug? I don't know; at least I know how to get it working now.

**Comment by Gama11** [comment]
Flash is the display configuration that's active by default. The "Code Completion" section of the flixel guide I linked explains how to switch between them.

The hxml files used by vshaxe are generated on build.

**Comment by ashes999** [comment]
Sorry, I didn't read it through in detail. I'm not a fan of how many steps it is, but at least it works. Thanks again.

**Comment by Gama11** [comment]
Agreed, the setup process could be a lot smoother. :)

**Comment by ashes999** [comment]
Guys, I cannot thank you enough for this. Really, this makes a huge difference in my development productivity. Thanks again for taking the time to help me sort through this!

**Comment by nightblade9** [comment]
I ran into exactly this issue today. AutoComplete doesn't work until I build `flash`, even though  `settings.json` contains lots of XML files:

```
{
	"haxe.displayConfigurations": [
		["export/flash/haxe/debug.hxml"],
		["export/flash/haxe/release.hxml"],
		["export/html5/haxe/debug.hxml"],
		["export/html5/haxe/release.hxml"],
		["export/windows/cpp/haxe/debug.hxml"],
		["export/windows/cpp/haxe/release.hxml"],
		["export/windows/neko/haxe/debug.hxml"],
		["export/windows/neko/haxe/release.hxml"],
		["export/linux/cpp/haxe/debug.hxml"],
		["export/linux/cpp/haxe/release.hxml"],
		["export/linux/neko/haxe/debug.hxml"],
		["export/linux/neko/haxe/release.hxml"],
		["export/mac/cpp/haxe/debug.hxml"],
		["export/mac/cpp/haxe/release.hxml"],
		["export/mac/neko/haxe/debug.hxml"],
		["export/mac/neko/haxe/release.hxml"],
		["export/android/haxe/debug.hxml"],
		["export/android/haxe/release.hxml"],
		["export/ios/haxe/debug.hxml"],
		["export/ios/haxe/release.hxml"]
	],
	"haxe.diagnosticsPathFilter": "${workspaceRoot}/source",
	"search.exclude": {
		"export/**/**.hx": true
	},
	"codedox": {
		"commentdescription": "\n",
		"paramFormat": "@param   ${name} ",
		"returnFormat": "@return  ${type}"
	}
}
```

@ashes999 did you ever find a way to make this work without building against the `flash` target first?

**Comment by RealyUniqueName** [comment]
@nightblade9 Try adding `"--cwd", "exprt/{target}/haxe"` to all of your display configurations.
Like htis:
```json
{
	"haxe.displayConfigurations": [
		["--cwd", "export/flash/haxe", "export/flash/haxe/debug.hxml"],		
		["--cwd", "export/windows/cpp/haxe", "export/windows/cpp/haxe/release.hxml"]
	]
}
```

**Comment by Gama11** [comment]
I think the working directory actually has to be the project root directory for OpenFL.

**Comment by nightblade9** [comment]
@Gama11 of course, the root directory in VSCode (and the location of my `.vscode`) is the root directory for OpenFL.

@RealyUniqueName looks promising but autocomplete just hangs and shows `Loading ...` forever.

**Comment by Gama11** [comment]
I meant for completion. `--cwd` changes the **c**urrent **w**orking **d**irectory.

**Comment by nightblade9** [comment]
> I meant for completion

Sorry, I don't understand what you  mean. Could you be a bit more verbose please?

**Comment by Gama11** [comment]
`haxe.displayConfigurations` is for code completion. The original configuration with `--cwd` seems correct to me.

**Comment by nightblade9** [comment]
Looks okay to me too, but unfortunately doesn't work -- VSCode gets stuck showing `Loading` for a really long time. (Full disclosure: I deleted `export`, built only `neko`, and didn't wait more than 10-20s to see if this ever finished.)

**Comment by DigitalNekkoDev** [comment]
We have the same problem too. My auto-completion seems loading forever
